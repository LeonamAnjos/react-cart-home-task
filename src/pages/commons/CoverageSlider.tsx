import React from "react";
import Slider from "@mui/material/Slider";

const stepSlider = 10;

type ProductCardDetailsProps = {
  initialCoverage?: number | undefined;
  minCoverage: number;
  maxCoverage: number;
  onCoverageChange?: (coverage: number) => void | undefined;
};

const CoverageSlider = ({
  initialCoverage,
  minCoverage,
  maxCoverage,
  onCoverageChange,
}: ProductCardDetailsProps) => {
  const [coverage, setCoverage] = React.useState<number>(
    initialCoverage ?? maxCoverage
  );

  const handleChange = (_: Event, newValue: number | number[]): void => {
    if (typeof newValue === "number") {
      setCoverage(newValue);
      onCoverageChange && onCoverageChange(newValue);
    }
  };

  return (
    <Slider
      value={coverage}
      min={minCoverage}
      max={maxCoverage}
      step={stepSlider}
      valueLabelDisplay="auto"
      onChange={handleChange}
    />
  );
};

export default CoverageSlider;
