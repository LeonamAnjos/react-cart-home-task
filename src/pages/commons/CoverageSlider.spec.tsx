import { render, fireEvent, screen } from "@testing-library/react";
import CoverageSlider, { CoverageSliderProps } from "./CoverageSlider";

describe("<CoverageSlider />", () => {
  const customRender = (props?: Partial<CoverageSliderProps>) => {
    return render(
      <CoverageSlider
        minCoverage={10}
        maxCoverage={90}
        initialCoverage={50}
        {...props}
      ></CoverageSlider>
    );
  };

  it("SHOULD render", () => {
    const { container } = customRender();

    expect(container).toMatchSnapshot();
  });

  it("SHOULD notify onCoverageChange event", async () => {
    const handleCoverageChange = jest.fn();

    const { container } = customRender({
      onCoverageChange: handleCoverageChange,
    });

    const element = screen.getByTestId("coverage-slider");
    await fireEvent.mouseDown(element, { clientX: 1 });

    expect(handleCoverageChange).toBeCalledTimes(1);
    expect(container).toMatchSnapshot();

    await fireEvent.mouseDown(element, { clientX: -1 });

    expect(handleCoverageChange).toBeCalledTimes(2);
    expect(container).toMatchSnapshot();
  });
});
