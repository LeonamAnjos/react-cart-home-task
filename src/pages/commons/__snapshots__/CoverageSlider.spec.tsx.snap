// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CoverageSlider /> SHOULD notify onCoverageChange event 1`] = `
<div>
  <span
    class="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root css-187mznn-MuiSlider-root"
    data-testid="coverage-slider"
  >
    <span
      class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
    />
    <span
      class="MuiSlider-track css-1gv0vcd-MuiSlider-track"
      style="left: 0%; width: 100%;"
    />
    <span
      class="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb Mui-active css-eg0mwd-MuiSlider-thumb"
      data-index="0"
      style="left: 100%;"
    >
      <input
        aria-orientation="horizontal"
        aria-valuemax="90"
        aria-valuemin="10"
        aria-valuenow="90"
        data-index="0"
        max="90"
        min="10"
        step="10"
        style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
        type="range"
        value="90"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-valueLabelOpen MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
      >
        <span
          class="MuiSlider-valueLabelCircle"
        >
          <span
            class="MuiSlider-valueLabelLabel"
          >
            90
          </span>
        </span>
      </span>
    </span>
  </span>
</div>
`;

exports[`<CoverageSlider /> SHOULD notify onCoverageChange event 2`] = `
<div>
  <span
    class="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root css-187mznn-MuiSlider-root"
    data-testid="coverage-slider"
  >
    <span
      class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
    />
    <span
      class="MuiSlider-track css-1gv0vcd-MuiSlider-track"
      style="left: 0%; width: 0%;"
    />
    <span
      class="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb Mui-active css-eg0mwd-MuiSlider-thumb"
      data-index="0"
      style="left: 0%;"
    >
      <input
        aria-orientation="horizontal"
        aria-valuemax="90"
        aria-valuemin="10"
        aria-valuenow="10"
        data-index="0"
        max="90"
        min="10"
        step="10"
        style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
        type="range"
        value="10"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-valueLabelOpen MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
      >
        <span
          class="MuiSlider-valueLabelCircle"
        >
          <span
            class="MuiSlider-valueLabelLabel"
          >
            10
          </span>
        </span>
      </span>
    </span>
  </span>
</div>
`;

exports[`<CoverageSlider /> SHOULD render 1`] = `
<div>
  <span
    class="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root css-187mznn-MuiSlider-root"
    data-testid="coverage-slider"
  >
    <span
      class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
    />
    <span
      class="MuiSlider-track css-1gv0vcd-MuiSlider-track"
      style="left: 0%; width: 50%;"
    />
    <span
      class="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
      data-index="0"
      style="left: 50%;"
    >
      <input
        aria-orientation="horizontal"
        aria-valuemax="90"
        aria-valuemin="10"
        aria-valuenow="50"
        data-index="0"
        max="90"
        min="10"
        step="10"
        style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
        type="range"
        value="50"
      />
      <span
        aria-hidden="true"
        class="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
      >
        <span
          class="MuiSlider-valueLabelCircle"
        >
          <span
            class="MuiSlider-valueLabelLabel"
          >
            50
          </span>
        </span>
      </span>
    </span>
  </span>
</div>
`;
